codegen/module.go

```diff
@@ -218,6 +235,145 @@ func appendUniqueModule(
 	return append(classDeps, instance)
 }
 
+func (system *ModuleSystem) populateRecursiveDependencies(
+	instances []*ModuleInstance,
+) error {
+	for _, classInstance := range instances {
+		recursiveDeps := map[string]map[string]*ModuleInstance{}
+		err := resolveRecursiveDependencies(
+			classInstance,
+			recursiveDeps,
+		)
+		if err != nil {
+			return err
+		}
+
+		for _, className := range system.classOrder {
+			moduleMap, ok := recursiveDeps[className]
+
+			if !ok {
+				continue
+			}
+
+			classInstance.DependencyOrder = append(
+				classInstance.DependencyOrder,
+				className,
+			)
+
+			moduleList := make([]*ModuleInstance, len(moduleMap))
+			index := 0
+			for _, moduleInstance := range moduleMap {
+				moduleList[index] = moduleInstance
+				index++
+			}
+
+			sortedModuleList, err := sortDependencyList(className, moduleList)
+			if err != nil {
+				return err
+			}
+
+			classInstance.RecursiveDependencies[className] = sortedModuleList
+		}
+	}
+
+	return nil
+}
+
+func resolveRecursiveDependencies(
+	instance *ModuleInstance,
+	resolvedDeps map[string]map[string]*ModuleInstance,
+) error {
+	for className, depList := range instance.ResolvedDependencies {
+		classDeps := resolvedDeps[className]
+
+		if classDeps == nil {
+			classDeps = map[string]*ModuleInstance{}
+			resolvedDeps[className] = classDeps
+		}
+
+		for _, dep := range depList {
+			if classDeps[dep.InstanceName] == nil {
+				classDeps[dep.InstanceName] = dep
+				err := resolveRecursiveDependencies(dep, resolvedDeps)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+type sortableDependencyList []*ModuleInstance
+
+func (s sortableDependencyList) Len() int {
+	return len(s)
+}
+
+func (s sortableDependencyList) Less(i int, j int) bool {
+	return s[i].InstanceName < s[j].InstanceName
+}
+
+func (s sortableDependencyList) Swap(i int, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func sortDependencyList(
+	className string,
+	instances []*ModuleInstance,
+) ([]*ModuleInstance, error) {
+	instanceList := sortableDependencyList(instances[:])
+	sort.Sort(instanceList)
+	sorted := make([]*ModuleInstance, len(instances))
+
+	for i, instance := range instances {
+		insertIndex := i
+		for j := 0; j < i; j++ {
+			if insertIndex == i {
+				if peerDepends(sorted[j], instance) {
+					insertIndex = j
+				}
+			}
+
+			if insertIndex != i {
+				if peerDepends(instance, sorted[j]) {
```

Comment:
It will be prevented as long as the peerDepends function looks at the recursively resolved dependencies (i.e. that is all dependencies in the dependency tree of the module) and that I am computing the recursive dependencies accurately (which might not be the case right now). Will add some tests for sure.